### PR TITLE
Use NCCL instead of ray for control-plane communication to remove serialization overhead

### PR DESCRIPTION
**Pull request by zhuohan123**
This PR modifies vLLM to use NCCL instead of ray for control-plane communication. The architectural change of vLLM with this PR can be summarized in the following figure:

image

Before this change, vLLM has one driver process (only on GPU), and N worker processes, each of which is a ray actor that manages one GPU. After this change, we will move one worker into the driver process (driver worker), and keep N-1 ray workers. All the control messages will be broadcast from the driver worker to all the remaining workers with NCCL. This avoids the high sterilization cost of ray communication.

For the throughput benchmark of LLaMA-70B on 8 A100-40G GPUs:

```
On ShareGPT Dataset
  Before this PR: 3.01 reqs/s
  With this PR: 5.08 reqs/s
With Batch size 512, input len 1031, output len 317
  Before this PR: 2.48 reqs/s
  With this PR: 3.48 reqs/s
```

This PR has passed most of the tests and is ready for review.

Should be merged after #2270, #2273.

### CODE REVIEW

File: vllm/model_executor/parallel_utils/communication_op.py
Comment (zhuohan123) [reply vllm/model_executor/parallel_utils/communication_op.py:52]: 😅 great catch. Will fix

Comment (hanzhi713) [reply vllm/model_executor/parallel_utils/communication_op.py:52]: Lucky catch. Browsed through the code in 10s.

Comment (Lvjinhong) [reply vllm/model_executor/parallel_utils/communication_op.py]: Hello, When can this branch be merged into the main branch, and will it bring significant performance improvements?

Comment (zhuohan123) [reply vllm/model_executor/parallel_utils/communication_op.py]: Hi @Lvjinhong Please find the performance numbers in the description of this PR. This PR is waiting for review and will be merged soon.

**Comment by esmeetu** [comment]
Hi @zhuohan123, this PR doesn't work for me.
`python -m vllm.entrypoints.openai.api_server --model /Llama-2-7b-chat-hf --tensor-parallel-size 4 --dtype half --enforce-eager`

```

  File "/home/roy/vllm/vllm/engine/llm_engine.py", line 599, in _process_model_outputs
    for seq_group, outputs in zip(scheduled_seq_groups, output):
TypeError: 'NoneType' object is not iterable
```

**Comment by zhuohan123** [comment]
> Hi @zhuohan123, this PR doesn't work for me. `python -m vllm.entrypoints.openai.api_server --model /Llama-2-7b-chat-hf --tensor-parallel-size 4 --dtype half --enforce-eager`
> 
> ```
> 
>   File "/home/roy/vllm/vllm/engine/llm_engine.py", line 599, in _process_model_outputs
>     for seq_group, outputs in zip(scheduled_seq_groups, output):
> TypeError: 'NoneType' object is not iterable
> ```

@esmeetu Sorry this is a bug! Can you test the latest commit again?

**Comment by esmeetu** [comment]
> > Hi @zhuohan123, this PR doesn't work for me. `python -m vllm.entrypoints.openai.api_server --model /Llama-2-7b-chat-hf --tensor-parallel-size 4 --dtype half --enforce-eager`
> 
> > 
> 
> > ```
> 
> > 
> 
> >   File "/home/roy/vllm/vllm/engine/llm_engine.py", line 599, in _process_model_outputs
> 
> >     for seq_group, outputs in zip(scheduled_seq_groups, output):
> 
> > TypeError: 'NoneType' object is not iterable
> 
> > ```
> 
> 
> 
> @esmeetu Sorry this is a bug! Can you test the latest commit again?

@zhuohan123 Ok now.👍

**Comment by esmeetu** [comment]
Hi, @zhuohan123, I tested this for many times, but why i didn't see any speed boost by using tp4 or tp8 on T4?🤔 I guess the bottleneck for me might be GPU FLOPS...

File: vllm/model_executor/layers/sampler.py
Comment (esmeetu) [inline vllm/model_executor/layers/sampler.py:48]: Why not putting this at the beginning of this method?

File: vllm/worker/model_runner.py
Comment (WoosukKwon) [inline vllm/worker/model_runner.py:390]: IIUC,
```suggestion
            if py_data["slot_mapping_size"] is not None:
```
Same for other places.

File: vllm/worker/model_runner.py
Comment (WoosukKwon) [inline vllm/worker/model_runner.py:347]: Could you add a simple comment explaining what we are doing here? e.g.,
```python
# Broadcast the input data. For input tensors, we first broadcast its shape.
```

File: vllm/model_executor/sampling_metadata.py
Comment (WoosukKwon) [inline vllm/model_executor/sampling_metadata.py:43]: Why is it -1 instead of 0?

File: vllm/model_executor/sampling_metadata.py
Comment (WoosukKwon) [inline vllm/model_executor/sampling_metadata.py:2]:
```suggestion
from typing import Dict, List, Optional, Tuple
```

File: vllm/engine/llm_engine.py
Comment (WoosukKwon) [inline vllm/engine/llm_engine.py:21]:
```suggestion
from vllm.utils import Counter, set_cuda_visible_devices, get_ip, get_open_port
```

File: vllm/engine/llm_engine.py
Comment (WoosukKwon) [inline vllm/engine/llm_engine.py:23]: Now that we don't import `ray.air`, can we remove `pandas` and `pyarrow` from vLLM's dependencies?

File: vllm/engine/llm_engine.py
Comment (WoosukKwon) [inline vllm/engine/llm_engine.py:141]:
```suggestion
            0,  # local_rank
            0,  # rank
```

File: vllm/engine/llm_engine.py
Comment (WoosukKwon) [inline vllm/engine/llm_engine.py:638]: Do we need this `if`?

File: vllm/engine/llm_engine.py
Comment (WoosukKwon) [inline vllm/engine/llm_engine.py:641]: Is this true? I thought only the driver worker returned the sampled outputs.

File: vllm/worker/worker.py
Comment (WoosukKwon) [inline vllm/worker/worker.py:190]: This pattern is repeated. Can we somehow factor this out?

**Review by WoosukKwon — COMMENTED** [review]
@zhuohan123 Thanks for the great work! I love the new architecture, and also like how you tackle the asymmetry between the driver and the workers. Maybe @Yard1 should also take a look at this PR, as it includes major changes in how we initiate/orchestrate Ray clusters.

File: vllm/model_executor/sampling_metadata.py
Comment (zhuohan123) [reply vllm/model_executor/sampling_metadata.py:43]: I think both works here. Let me change it to 0.

File: vllm/worker/worker.py
Comment (zhuohan123) [reply vllm/worker/worker.py:187]: Added a comment to explain 7

File: vllm/worker/worker.py
Comment (zhuohan123) [reply vllm/worker/worker.py]: Do you mean factoring the following part out?

```python
if self.is_driver_worker:
    # some manual serialization logic to get the python object and 
    # the tensors to broadcast
    ...
    # broadcast the python object
    torch.distributed.broadcast_object_list([python_object], src=0)
    # broadcast the tensors
    torch.distributed.broadcast(..., src=0)
else:
    # receive the python object
    python_object = torch.distributed.broadcast_object_list(
        [None], src=0)[0]
    # receive the tensors
    torch.distributed.broadcast(..., src=0)
```

I can give it a try, but the potential issue is that the manual serialization logic can be pretty arbitrary.

File: vllm/model_executor/layers/sampler.py
Comment (esmeetu) [reply vllm/model_executor/layers/sampler.py:51]: Is logits computation a process of sampling? It seems a middle state process between model and sampling. 🤔

**Comment by sfc-gh-hazhang** [comment]
ETA?

**Comment by zhisbug** [comment]
ETA? Let me know when I can run some test

@WoosukKwon we can tag a new ver after this one to merge

**Review by njhill — COMMENTED** [review]
Looks like a very nice speed-up!

**Comment by oliviadzy** [comment]
hello!I'm very interested in your idea and tried running it on my computer, but I encountered a 'GPU out of memory' error. However, I didn't encounter this issue when using the original vllm inference framework. Could you please tell me what might be the problem?
屏幕截图 2023-12-31 163007

**Comment by Lccwr** [comment]
Hello, encountered some problems while testing on your branch. 
python benchmark_latency.py  --model llama-70B-hf --tensor-parallel-size 4 --input-len 1536 --output-len 512 --batch-size 120 --num-iters 2 --enforce-eager
721D656E-889C-4B9E-9B16-3816980EB20F

File: vllm/model_executor/layers/sampler.py
Comment (WoosukKwon) [reply vllm/model_executor/layers/sampler.py:46]: I think this is a very good point. @zhuohan123 Could you make the change accordingly?

File: vllm/model_executor/layers/sampler.py
Comment (WoosukKwon) [reply vllm/model_executor/layers/sampler.py:51]: I kinda agree with @esmeetu. Maybe we should make it a separate stage. But let's discuss it in a future PR.

File: vllm/worker/model_runner.py
Comment (WoosukKwon) [reply vllm/worker/model_runner.py:379]: I think @njhill's point makes a lot of sense. Can we leave a TODO comment (or a issue) about it?

File: vllm/engine/ray_utils.py
Comment (WoosukKwon) [inline vllm/engine/ray_utils.py:37]:
```suggestion
        def set_cuda_visible_devices(self, device_ids) -> None:
```

File: vllm/engine/ray_utils.py
Comment (WoosukKwon) [inline vllm/engine/ray_utils.py:32]: nit: please annotate the return type.
```suggestion
        def get_node_and_gpu_ids(self) -> Tuple[int, List[int]]:
```

**Review by WoosukKwon — APPROVED** [review]
@zhuohan123 Awesome! Thanks for addressing my comments. Let's merge this asap after addressing the remaining minor comments from @njhill!

File: vllm/model_executor/layers/sampler.py
Comment (zhuohan123) [reply vllm/model_executor/layers/sampler.py:51]: Added a TODO here

File: vllm/model_executor/layers/sampler.py
Comment (zhuohan123) [reply vllm/model_executor/layers/sampler.py:46]: Thanks for the suggestion! Changed in the latest commit.

File: vllm/worker/model_runner.py
Comment (zhuohan123) [reply vllm/worker/model_runner.py:379]: Added a TODO.

File: vllm/engine/llm_engine.py
Comment (Yard1) [inline vllm/engine/llm_engine.py:168]: I don't think this is a good idea as it will prevent using vLLM on clusters where the head node is CPU only, or with Ray Serve. I would instead suggest the following:
1. start one dummy worker for each bundle in the placement group
2. get node IPs of all the dummy workers
3. see if one of the dummy workers matches the node IP of the current process
4. if not, throw exception, otherwise, this dummy worker will be the "real" dummy worker
5. tear down all dummy workers and start real workers + the "real" dummy worker

this approach should be correct under any circumstance and allow for multiple instances of vLLM to correctly follow Ray scheduling contracts.

**Comment by Yard1** [comment]
@zhuohan123 Can you give me a few hours to look over this? Thanks!

File: vllm/engine/llm_engine.py
Comment (zhuohan123) [inline vllm/engine/llm_engine.py:176]: @Yard1 This is where we use the bundle with the `node:__internal_head__` resource to hold the resource for the driver worker.

**Comment by zhuohan123** [comment]
> @zhuohan123 Can you give me a few hours to look over this? Thanks!

Sure! Just highlighted several places where we changed the logic of how we use placement groups, which I think should be important for you to take a look.

File: vllm/model_executor/parallel_utils/communication_op.py
Comment (zhuohan123) [reply vllm/model_executor/parallel_utils/communication_op.py]: great catch! fixed.

File: vllm/engine/ray_utils.py
Comment (Yard1) [inline vllm/engine/ray_utils.py:120]: ditto, see comment above

File: vllm/worker/worker.py
Comment (zhuohan123) [reply vllm/worker/worker.py:170]: You are right. This change is not necessary for this PR. 

I originally changed the input arguments to cache operations to make serialization faster. After a second thought, I believe this is unnecessary and I will revert the changes of cache operations to make this PR simpler.

**Review by Yard1 — CHANGES_REQUESTED** [review]

File: vllm/engine/llm_engine.py
Comment (zhuohan123) [reply vllm/engine/llm_engine.py:168]: iiuc, you would like to avoid the requirement of using `node:__internal_head__`. Instead, we can just check the IP address of each worker, if a worker's IP matches driver's IP, then set that worker to be driver's dummy worker. If no worker's IP matches the driver's IP, we throw an exception.

Is my understanding correct? If so, I can change the code accordingly.

Comment (Yard1) [reply vllm/engine/llm_engine.py:168]: Yep, that's exactly it!

Comment (zhuohan123) [reply vllm/engine/llm_engine.py]: @Yard1 Fixed. Can you take a look again?

File: vllm/engine/llm_engine.py
Comment (Yard1) [inline vllm/engine/llm_engine.py:174]: As a minor optimization you can make it another loop so that the Workers can be initialized in a non-blocking fashion but considering that there's nothing really happening in the `__init__` I think  it's ok to leave it in (though it is an anti-pattern).

**Review by Yard1 — APPROVED** [review]
thanks, looks good!

**Pull request merged**
